extension-manifest-v3/src/utils/dnr-converter.js

```diff
@@ -9,45 +9,114 @@
-  });
+    const iframe = document.createElement('iframe');
+    iframe.setAttribute('src', 'https://ghostery.github.io/urlfilter2dnr/');
+    iframe.setAttribute('style', 'display: none;');
 
-  return pending;
+    creating = new Promise((resolve) => {
+      iframe.addEventListener('load', () => resolve(iframe));
+      document.head.appendChild(iframe);
+    });
+
+    return creating;
+  }
+
+  let requestCount = 0;
+
+  async function convert(filter) {
+    const iframe = await createIframe();
+    const requestId = `request${requestCount++}`;
+
+    iframe.contentWindow.postMessage(
+      {
+        action: 'convert',
+        converter: 'adguard',
+        filters: [requestId, filter],
+      },
+      '*',
+    );
+
+    return new Promise((resolve) => {
+      requests.set(requestId, resolve);
+    });
+  }
+
+  return convert;
 }
 
-let requestCount = 0;
-export async function convert(filter) {
-  const iframe = await createIframe();
-  const requestId = `request${requestCount++}`;
-
-  iframe.contentWindow.postMessage(
-    {
-      action: 'convert',
-      converter: 'adguard',
-      filters: [requestId, filter],
-    },
-    '*',
-  );
-
-  return new Promise((resolve) => {
-    requests.set(requestId, resolve);
-  });
+function createMV3Converter() {
+  async function setupOffscreenDocument() {
+    const path = 'pages/offscreen/urlfilter2dnr/index.html';
+    const offscreenUrl = chrome.runtime.getURL(path);
+    const existingContexts = await chrome.runtime.getContexts({
+      contextTypes: ['OFFSCREEN_DOCUMENT'],
+      documentUrls: [offscreenUrl],
+    });
+
+    if (existingContexts.length > 0) {
+      return;
+    }
+
+    if (creating) {
+      await creating;
+    } else {
+      creating = chrome.offscreen.createDocument({
+        url: path,
+        reasons: [chrome.offscreen.Reason.IFRAME_SCRIPTING],
+        justification:
+          'Convert network filters to DeclarativeNetRequest format.',
+      });
+      await creating;
+      creating = null;
+    }
+  }
+
+  async function createOffscreenDocument() {
+    async function hasPermission() {
+      return chrome.permissions.contains({
+        permissions: ['offscreen'],
+      });
+    }
+    if (!(await hasPermission())) {
+      await chrome.permissions.request({ permissions: ['offscreen'] });
+      if (!(await hasPermission())) {
+        throw new Error('Ghostery requires "offscreen" permission');
+      }
+    }
+    await setupOffscreenDocument();
+  }
+
+  async function convert(filter) {
+    try {
+      await createOffscreenDocument();
+    } catch (e) {
+      return { errors: [e.message], rules: [] };
+    }
+    return (
+      (await chrome.runtime.sendMessage({
+        action: 'offscreen:urlfitler2dnr:convert',
+        filter,
+      })) || { errors: ['failed to innitiate offscreen document'], rules: [] }
+    );
+  }
+
+  return convert;
 }
+
+export default chrome && chrome.runtime.getManifest?.().manifest_version === 3
```

Comment:
The idea is perfect, but it should be done on the build step, and it is not strictly related to the `manifest_version`, rather the platform. Only chromium has a permission to `offscreen` so I suppose, that only it supports it. The Safari supports version 3, which I would like to migrate to (and I am not sure if you will have offscreen document there).


```suggestion
export default __PLATFORM__ !== 'safari' && __PLATFORM__ !== 'firefox'
  ? createOffscreenConverter()
  : createDocumentConverter()
```

Mind the new function naming - now they are not related to mv... rather what they actually do...

Reply:
cannot be done on a build as this is used inside of the offscreen document, but I will update the check

Reply:
It does not matter, platform is a primary condition, and it's always available.. and it's a build time check.